checkpoint/recovery.go

```diff
@@ -108,6 +108,13 @@ func Recover(
 	fs afero.Fs,
 	cfg *RecoverConfig,
 ) (*PreservedData, error) {
+	if len(cfg.Uri) == 0 {
+		return nil, nil
```

Comment:
I'm not sure that returning no error here is even correct? 🤔 shouldn't it return an error if no URI was passed?

Reply:
I intentionally moved this check from node.go to be able to test it (testing the `App` is unpleasant). Empty URI essentially stands for "checkpoint-restore" disabled and it's a valid value, hence returning `nil` error.

> I think that methods that return (*pointer,err) shouldn't return nil, nil as it can cause panics on the calling context.

Well, it _depends_. Unfortunately, Go doesn't have something like Rust `Option` and we need to escape to pointers with `nil` equivalent to `None` :(. I could document the possible return values if that would help, although I'd prefer to come up with an API that's hard to misuse rather than depend on comments which are rarely read.

Reply:
The caller could just check if `URI` is not set and then not call the function? as the function would return an error otherwise.

Reply:
> The caller could just check if `URI` is not set and then not call the function? as the function would return an error otherwise.

Then how do you propose to test that an empty URI is a valid node configuration value?

Reply:
If its empty it won't call the `Recover` function? A negative doesn't need to be tested?

Reply:
It's also indirectly tested by every other test in the `node` package which would fail if `Recover` would do something on a non-empty `URI` 😉